### PR TITLE
Cache resources in SchedulingQueue

### PR DESCRIPTION
**Pull request by pcmoritz**
## Related issue number

https://github.com/ray-project/ray/issues/3193

### CODE REVIEW

**Comment by pcmoritz** [comment]
Jenkins retest this please

**Review by stephanie-wang — CHANGES_REQUESTED** [review]
Instead of doing the bookkeeping in separate methods (like `AddReadyQueueResources` and `RemoveReadyQueueResources`), how about we add a member `current_resource_load_` to `TaskQueue`, and then update it whenever `TaskQueue::AppendTask` or `TaskQueue::RemoveTask` is called? And then we can implement `SchedulingQueue::GetResourceLoad()` to just return `ready_tasks_.GetResourceLoad()`?

File: src/ray/raylet/scheduling_queue.cc [hunk 1]
```diff
@@ -80,30 +80,25 @@ bool SchedulingQueue::TaskQueue::AppendTask(const TaskID &task_id, const Task &t
...
-  task_map_.erase(task_found_iterator);
-  task_list_.erase(list_iterator);
+  // Resource bookkeeping
+  current_resource_load_.AddResources(task.GetTaskSpecification().GetRequiredResources());
   return true;
 }
 
 bool SchedulingQueue::TaskQueue::RemoveTask(const TaskID &task_id,
-                                            std::vector<Task> &removed_tasks) {
+                                            std::vector<Task> *removed_tasks) {
```
Comment (stephanie-wang) [inline src/ray/raylet/scheduling_queue.cc]: Why this change? We should avoid using raw pointers. If the idea is to share code with the other `RemoveTask`, let's just pass in a temporary `std::vectorTask>` there.

Comment (pcmoritz) [reply src/ray/raylet/scheduling_queue.cc]: This is the generic rule for arguments but the specific one for out-arguments is pointers:

```
In fact it is a very strong convention in Google code that [...] output arguments are pointers.
```

The convention is clear and we decided to follow it (I know that in other parts of the code we use non-const references for this, and we should change that I think).

It also avoids creating the temporary vector in this case.

**Review by stephanie-wang — APPROVED** [review]
